stats/thresholds_parser.go

```diff
@@ -0,0 +1,230 @@
+		ParseAggregationMethod(),
+		c.DiscardAll(c.Whitespace()),
+		ParseOperator(),
+		c.DiscardAll(c.Whitespace()),
+		ParseValue(),
+		c.DiscardAll(c.Newline()),
+	)
+
+	return func(input []rune) c.Result {
+		res := parser(input)
+		if res.Err != nil {
+			return res
+		}
+
+		return c.Success(res.Payload, res.Remaining)
+	}
+}
+
+// ParseAggregationMethod parses any aggregation method as defined in
+// the BNF: `aggregation_method -> trend | rate | gauge | counter`.
+// The Result's `Payload interface{}` value will hold the
+// aggregation method name as a string.
+func ParseAggregationMethod() c.Parser {
+	parser := c.Expect(c.Alternative(
+		ParseCounter(),
+		ParseGauge(),
+		ParseRate(),
+		ParseTrend(),
+		ParsePercentile(),
+	), "aggregation method")
+
+	return func(input []rune) c.Result {
+		res := parser(input)
+		if res.Err != nil {
+			return res
+		}
+
+		return c.Success(res.Payload, res.Remaining)
+	}
+}
+
+// ParseOperator parses a threshold expression supported operator
+// as defined in the BNF: `operator -> ">" | ">=" | "<=" | "<" | "==" | "===" | "!="`.
+// The Result's `Payload interface{}` value will hold the
+// operator expression as a string.
+func ParseOperator() c.Parser {
+	parser := c.Expect(c.Alternative(
+		c.Tag(">="),
+		c.Tag("<="),
+		c.Tag(">"),
+		c.Tag("<"),
+		c.Tag("==="),
+		c.Tag("=="),
+		c.Tag("!="),
+	), "operator")
+
+	return func(input []rune) c.Result {
+		res := parser(input)
+		if res.Err != nil {
+			return res
+		}
+
+		return c.Success(res.Payload.(string), res.Remaining)
+	}
+}
+
+// ParseTrend parses a trend aggregation method as defined in
+// the BNF: `trend -> "avg" | "min" | "max" | "med" | percentile`.
+// The Result's `Payload interface{}` value will hold the
+// trend's aggregation method name as a string.
+func ParseTrend() c.Parser {
+	parser := c.Expect(c.Alternative(
+		c.Tag("mean"),
+		c.Tag("min"),
+		c.Tag("max"),
+		c.Tag("avg"),
+		c.Tag("med"),
+		ParsePercentile(),
+	), "trend aggregation method")
+
+	return func(input []rune) c.Result {
+		res := parser(input)
+		if res.Err != nil {
+			return res
+		}
+
+		return c.Success(res.Payload, res.Remaining)
+	}
+}
+
+// ParsePercentile parses a percentile as defined in
+// the BNF: `percentile -> "p(" float ")"`. The Result's `Payload interface{}`
+// value will hold the percentile expression as a string.
+func ParsePercentile() c.Parser {
+	parser := c.Expect(c.Sequence(
+		c.Tag("p("),
+		c.Float(),
+		c.Char(')'),
+	), "percentile")
+
+	return func(input []rune) c.Result {
+		res := parser(input)
+		if res.Err != nil {
+			return res
+		}
+
+		parsed, ok := res.Payload.([]interface{})
+		if !ok {
+			err := fmt.Errorf("failed parsing percentile expression; " +
+				"reason: converting ParsePercentile() parser result's payload to []interface{} failed",
+			)
+			res.Err = c.NewFatalError(input, err, "percentile")
+		}
+
+		percentile := fmt.Sprintf("%s%g%s", parsed[0].(string), parsed[1].(float64), parsed[2].(string))
```

Comment:
That's a very good point, I do not check that the percentile is with bounds `[0, 100]` at the moment. Going to add that + a test for it.

On the front of precision, the `%g` formatter in the `Sprintf` call ensures to limit the decimal precision to the minimum necessary. For instance, `99.0` would be represented as `99`, but `99.1` would remain `99.1` while `99.10` would be truncated to `99.1`.

Reply:
Ah, I'm sorry, I should have been more verbose in this comment about precision :sweat_smile: I meant not the precision in this particular line but in all steps of thresholds processing together: string -> float64 -> string. I.e. it's worth to consider if there can be any loss / misrepresentation at some point and if it can, IMO, it should be either fixed or documented. I can't say that I see the problem with conversions right now but I'd have run some additional tests for edge cases, etc., to be sure. Just a suggestion :slightly_smiling_face: 

However, this does work with current `master`:
```
http_req_failed: ['rate<123e5'],
http_req_duration: ['p(34)<123e-5'],
```
But I don't think this would work in this PR because exponent is not processed in the value. To be clear, I'm not saying this _must_ work... e.g. was anyone actually using this form of floats in their scripts? Maybe it can be dropped? But this is probably something to decide / investigate more.